src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -2872,6 +2882,40 @@ private int[] SerializeMethodBodies(BlobBuilder ilBuilder, PdbWriter nativePdbWr
             return bodyOffsets;
         }
 
+        private int SerializeThrowNullMethodBody(MethodBodyStreamEncoder encoder)
+        {
+            var il = ThrowNullIL;
+
+            // Don't do small body method caching during deterministic builds until this issue is fixed
+            // https://github.com/dotnet/roslyn/issues/7595
+            int bodyOffset;
+            if (!_deterministic && _smallMethodBodies.TryGetValue(il, out bodyOffset))
+            {
+                return bodyOffset;
+            }
+
+            var encodedBody = encoder.AddMethodBody(
+                codeSize: il.Length,
+                maxStack: ThrowNullMaxStack,
+                exceptionRegionCount: 0,
+                hasSmallExceptionRegions: false);
+
+            // Don't do small body method caching during deterministic builds until this issue is fixed
+            // https://github.com/dotnet/roslyn/issues/7595
+            if (!_deterministic)
+            {
+                _smallMethodBodies.Add(il, encodedBody.Offset);
+            }
+
+            UserStringHandle mvidStringHandle = default(UserStringHandle);
+            Blob mvidStringFixup = default(Blob);
+            WriteInstructions(encodedBody.Instructions, il, ref mvidStringHandle, ref mvidStringFixup);
+
+            Debug.Assert(mvidStringHandle.Equals(default(UserStringHandle)) && mvidStringFixup.Equals(default(Blob)));
+
+            return encodedBody.Offset;
+        }
```

Comment:
Consider extracting a helper method from `SerializeMethodBody` that both `SerializeMethodBody` and `SerializeThrowNullMethodBody` can call:
```
SerializeMethodBody(
    MethodBodyStreamEncoder encoder,
    ImmutableArray<byte> il,
    uint maxStack,
    ImmutableArray<ExceptionRegion> exceptionRegions,
    StandaloneSignatureHandle localSignatureHandle)
```